src/documents/migrations/1053_document_pages_count.py

```diff
@@ -0,0 +1,109 @@
+# Generated by Django 4.2.16 on 2024-09-21 15:44
+
+import datetime
+from pathlib import Path
+
+import pikepdf
+from django.conf import settings
+from django.db import migrations
+from django.db import models
+from django.utils import timezone
+from django.utils.termcolors import colorize as colourise
+
+from documents.parsers import get_default_file_extension
+
+
+class Document:
+    """
+    Django's migrations restrict access to model methods, so this is a snapshot
+    of the methods that existed at the time this migration was written, since
+    we need to make use of a lot of these shortcuts here.
+    """
+
+    def __init__(self, doc):
+        self.pk = doc.pk
+        self.correspondent = doc.correspondent
+        self.title = doc.title
+        self.mime_type = doc.mime_type
+        self.filename = doc.filename
+        self.created = doc.created
+
+    def __str__(self) -> str:
+        # Convert UTC database time to local time
+        created = datetime.date.isoformat(timezone.localdate(self.created))
+
+        res = f"{created}"
+
+        if self.correspondent:
+            res += f" {self.correspondent}"
+        if self.title:
+            res += f" {self.title}"
+        return res
+
+    @property
+    def file_type(self):
+        return get_default_file_extension(self.mime_type)
+
+    @property
+    def source_path(self) -> Path:
+        if self.filename:
+            fname = str(self.filename)
+        return (settings.ORIGINALS_DIR / Path(fname)).resolve()
+
+
+def add_number_of_pages_to_pages_count(apps, schema_editor):
+    documentModel = apps.get_model("documents", "Document")
+
+    if not documentModel.objects.all().exists():
+        return
+
+    for doc in documentModel.objects.filter(mime_type="application/pdf"):
+        document = Document(doc)
+
+        print(
+            "    {} {} {}".format(
+                colourise("*", fg="green"),
+                colourise("Calculating number of pages for", fg="white"),
+                colourise(document.filename, fg="cyan"),
+            ),
+        )
+
+        pdf = pikepdf.open(document.source_path)
```

Comment:
Wonder if this should have try/ catch like the other spot

Reply:
Like this ?
```
try:
    pdf = pikepdf.open(document_path)
except pikepdf.PdfError as e:
    self.log.warning(f"Error opening PDF file: {e}")
except Exception as e:
    self.log.warning(f"An unexpected error occurred: {e}")
```

Or just add "with" like this ?
```
with pikepdf.open(document.source_path) as pdf:
```

I see error handling in other cases for a file operation, but not just for opening.

Reply:
I can dig into it a bit, I'm not sure if pikepdf does ever throw an error on open / pages, but yea I meant potentially wrapping the whole thing just in case. But yea, if that never throws an error agree wouldn't be needed, I do sort of assume it's possible

Reply:
Yea, I think this should get try/catch handling. Below is just trying on my local dev setup where the files get messed up frequently (which obviously does happen in the real world too):

```
    * Calculating number of pages for subdir_new_edited/2024/bug-test-workflows.pdf
Traceback (most recent call last):
  File "/src/manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/.venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/.venv/lib/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/.venv/lib/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/.venv/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/.venv/lib/python3.9/site-packages/django/core/management/base.py", line 106, in wrapper
    res = handle_func(*args, **kwargs)
  File "/.venv/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 356, in handle
    post_migrate_state = executor.migrate(
  File "/.venv/lib/python3.9/site-packages/django/db/migrations/executor.py", line 135, in migrate
    state = self._migrate_all_forwards(
  File "/.venv/lib/python3.9/site-packages/django/db/migrations/executor.py", line 167, in _migrate_all_forwards
    state = self.apply_migration(
  File "/.venv/lib/python3.9/site-packages/django/db/migrations/executor.py", line 252, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/.venv/lib/python3.9/site-packages/django/db/migrations/migration.py", line 132, in apply
    operation.database_forwards(
  File "/.venv/lib/python3.9/site-packages/django/db/migrations/operations/special.py", line 193, in database_forwards
    self.code(from_state.apps, schema_editor)
  File "/src/documents/migrations/1053_document_pages_count.py", line 71, in add_number_of_pages_to_pages_count
    pdf = pikepdf.open(document.source_path)
  File "/.venv/lib/python3.9/site-packages/pikepdf/_methods.py", line 392, in open
    stream = open(filename_or_stream, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: '/media/documents/originals/subdir_new_edited/2024/bug-test-workflows.pdf'
```

This completely halts the migration and leaves it in a state where it can't proceed (without fixing the issue). So I think best would be to just move on in these cases, log it. I've made a suggestion.